pkg/cloudprovider/aws/fleet/packing/packing.go

```diff
@@ -38,19 +42,141 @@ type Packings struct {
 	InstanceTypeOptions []string
 }
 
-func NewPacker(ec2 ec2iface.EC2API) *PodPacker {
-	return &PodPacker{ec2: ec2}
+// NewPacker returns a Packer implementation
+func NewPacker(ec2 ec2iface.EC2API) Packer {
+	return &podPacker{ec2: ec2}
 }
 
 // Pack returns the packings for the provided pods. Computes a set of viable
 // instance types for each packing of pods. Instance variety enables EC2 to make
-// better cost and availability decisions.
-func (p *PodPacker) Pack(ctx context.Context, pods []*v1.Pod) ([]*Packings, error) {
-	zap.S().Debugf("Successfully packed %d pod(s) onto %d node(s)", len(pods), 1)
-	return []*Packings{
-		{
-			InstanceTypeOptions: []string{"m5.large"}, // TODO, prioritize possible instance types
-			Pods:                pods,
-		},
-	}, nil
+// better cost and availability decisions. Pods provided are all schedulable in
+// the same zone as tightly as possible. It follows the First Fit Decreasing bin
+// packing technique, reference-
+// https://en.wikipedia.org/wiki/Bin_packing_problem#First_Fit_Decreasing_(FFD)
+func (p *podPacker) Pack(ctx context.Context, pods []*v1.Pod) ([]*Packings, error) {
+	// 1. Arrange pods in decreasing order by the amount of CPU requested, if
+	// CPU requested is equal compare memory requested.
+	sort.Sort(sort.Reverse(byResourceRequested{pods}))
+	return p.packSortedPods(pods)
+}
+
+// takes a list of pods sorted based on their resource requirements compared by CPU and memory.
+func (p *podPacker) packSortedPods(pods []*v1.Pod) ([]*Packings, error) {
+	// start with the smallest instance type and the biggest pod check how many
+	// pods can we fit, go to the next bigger type and check if we can fit more
+	// pods. Compare pods packed on all types and select the instance type with
+	// highest pod count.
+	estimator := &packingEstimator{isPodPacked: map[*v1.Pod]bool{}}
+	packings := []*Packings{}
+	for _, pod := range pods {
+		if estimator.isPodPacked[pod] {
+			continue
+		}
+		podsPacked, instances := estimator.calculatePackingForAllTypes(pods)
+		// checked all instance type and found no packing option
+		if len(podsPacked) == 0 || len(instances) == 0 {
+			return nil, fmt.Errorf("no capacity type found for packing %d pods", len(pods))
+		}
+		// keep a track of pods we have packed
+		for _, pod := range podsPacked {
+			estimator.isPodPacked[pod] = true
+		}
+		zap.S().Debugf("For %d pod(s) instance types selected are %v", len(podsPacked), instances)
+		packings = append(packings, &Packings{podsPacked, instances})
+	}
+	return packings, nil
+}
+
+type packingEstimator struct {
+	isPodPacked map[*v1.Pod]bool
+}
+
+func (p *packingEstimator) calculatePackingForAllTypes(pods []*v1.Pod) ([]*v1.Pod, []string) {
+	previousPacking := []*v1.Pod{}
```

Comment:
I'd avoid calling this previous "Packing" because there's a `Packing` struct type defined here, so it's confusing for a slice of `Pod`s to be called that.